django-hosting-heroku/README.md

```diff
@@ -0,0 +1,44 @@
+# How to Host Your Django Project on Heroku
+
+This is a companion project to a tutorial on Real Python. It's a bare-bones Django project without any views.
+
+## Instructions
+
+You should follow the step-by-step instructions in the tutorial to deploy this project to Heroku. Alternatively, you can run the finished project by following the instructions below.
+
+Download the code by cloning the `materials` repository:
+
+```shell
+$ git clone git@github.com:realpython/materials.git
+```
+
+Create a new local Git repository:
+
+```shell
+$ cd materials/django-hosting-heroku/source_code_final/portfolio-project
+$ git init
+$ git add .
+$ git commit -m "Initial commit"
+```
+
+Deploy the code to Heroku:
+
+```shell
+$ heroku login
+$ heroku create
+$ heroku config:set DEBUG=True SECRET_KEY=$(date | base64)
+$ git push heroku master
+$ heroku open
+```
+
+Optionally, for local development:
+
+```shell
+$ python3 -m venv .venv
+$ source .venv/bin/activate
+$ python3 -m pip install -r requirements.txt
+$ heroku config -s > .env
+$ heroku local
```

Comment:
🔄🤝 **Hand holding:** If your readers follow the steps as outlined in your `README.md` file exactly, they'll run into an error when attempting to serve their apps locally using `heroku local` in the final two steps where you also make them deploy to heroku:

```
heroku local
[OKAY] Loaded ENV .env File as KEY=VALUE Format
3:45:47 PM web.1 |  Watching for file changes with StatReloader
3:45:47 PM web.1 |  Exception in thread django-main-thread:
3:45:47 PM web.1 |  Traceback (most recent call last):
3:45:47 PM web.1 |    File "/Users/martin/Documents/real_python/materials/django-hosting-heroku/source_code_step_7/portfolio-project/.venv/lib/python3.9/site-packages/django/db/backends/postgresql/base.py", line 25, in <module>
3:45:47 PM web.1 |      import psycopg2 as Database
3:45:47 PM web.1 |    File "/Users/martin/Documents/real_python/materials/django-hosting-heroku/source_code_step_7/portfolio-project/.venv/lib/python3.9/site-packages/psycopg2/__init__.py", line 51, in <module>

-- snip --

3:45:47 PM web.1 |    Expected in: flat namespace
3:45:47 PM web.1 |   in /Users/martin/Documents/real_python/materials/django-hosting-heroku/source_code_step_7/portfolio-project/.venv/lib/python3.9/site-packages/psycopg2/_psycopg.cpython-39-darwin.so
```

The reason for this is the `DATABASE_URL` variable that gets copied into the `.env` file with the command:

```bash
heroku config -s > .env
```

They'll have to delete the `DATABASE_URL` variable from their `.env` file before running the local server with `heroku local` for everything to work as expected.

Can you add the necessary instructions to your `README.md` file to clarify this?

Reply:
This is might also be helpful: https://humberto.io/blog/tldr-generate-django-secret-key/

Nice one-liner for generating a random, URL-safe secret string with Python:

```console
$ python -c "import secrets; print(secrets.token_urlsafe())"
WGnlh-ZDPSNok6avqOiLWqHEVjl0r0ilpWivnODI-E4
```

Reply:
I'm not able to reproduce this :thinking:

Reply:
This might be caused by the `.env` files ignored by Git, which I had to force add.

Reply:
I did a fresh pull to go over the whole process again and I run into exactly the same issue @bzaczynski .

Just to clarify: if you go over the outlined steps from top to bottom, including:

- first deploying to heroku remotely
- then saving the heroku config to a `.env` file
- then running `heroku local`

You don't run into any errors?

I think it's because the remote deployment creates a `DATABASE_URL` variable for the remote PostgreSQL instance, and when you write that one to the local heroku config through the `.env` file, it can't properly operate:

```
DATABASE_URL='postgres://blnywyuwycvnpq:blablabla@ec2-52-1-20-236.compute-1.amazonaws.com:5432/d430esdevbvq52'
DEBUG=True
SECRET_KEY='9elXvmXLPBTZjC8RqOyWogfHRBs2pZlxtIE+HL9CT88=
```

Once I delete that variable, it works fine for me:

https://user-images.githubusercontent.com/12371494/124306057-2fcc5700-db66-11eb-9959-0cb3951d79ed.mov

Reply:
It works flawlessly for me. I googled your error message and it seems to be, surprise surprise, a Mac-specific problem. I'll add a note in the README.